### DISCUSSION TITLE
Discord bot cannot send messages with data from the api

### DISCUSSION THREAD

**Discussion by 0J33** [theOehrly/Fast-F1#267] [Q&A] [answered]
I'm working on a discord bot that uses the api to send text messages with things like fastest laps of a session, race results, season schedule, session event.

I got it to work with fastest laps but I can't seem to make it work with race results. I haven't tried schedule or event yet but I think it might have the same problem.

I know where the problem is but I can't figure out how to fix it.

This is the code for the fastest laps. It works fine. (yr is the year, rc is the grand prix, sn is the session)
<img width="714" alt="fastest laps" src="https://user-images.githubusercontent.com/90639992/193093018-4044ef4a-8511-4164-bc10-6e34b41c7be6.png">
<img width="335" alt="fl" src="https://user-images.githubusercontent.com/90639992/193095513-85de62b7-285b-4d04-bc8d-edfc3ea16637.png">
This is the message it send.

This is the code for the race results. It doesn't send a message with the race results and I get the message from the exception of the display error.
<img width="426" alt="results" src="https://user-images.githubusercontent.com/90639992/193093011-bb4791ac-771e-4357-929d-4ccbfaca276e.png">
<img width="258" alt="r" src="https://user-images.githubusercontent.com/90639992/193095757-a6fb2420-b839-48f3-b81d-3266eec1aaf2.png">
I know it got the data correctly because when I try to print the string msg in the console I get the race results, but it doesn't work in discord. I think the problem is in the line where I send the the discord message with the string msg, and I think it has to do with this line
`msg = f" {session.results}"` 
because in the fastest laps code it only works and sends the message when I use this line 
`msg = f" {fastest_laps[['Driver', 'LapTime', 'LapTimeDelta']]}"`
so I think I just need a different way to set the string msg but I'm not sure how to do it.

**Comment by theOehrly** [comment, marked as answer]
Well, this is very likely not a problem with FastF1. The objects that you are trying to print are Pandas DataFrames underneath. Some part of the conversion to string is obviously not working. 
The error message is not really helpful. Instead of just "throwing away" the information that the exception provides, you should log the exception. You can try something along the lines of 

```
try: 
    await message.channel.send(msg)
except Exception as exc:
    print(exc)
    # or
    logging.exception(exc)
    # or you could send the exception through slack as well
    await message.channel.send(str(exc))
```

Hopefully that will give you a clue as to what the actual problem is. Of course, you could also just run this in a debugger to check.

For completeness’ sake, there is an implementation difference between the `__repr__` method of `Laps` and `SessionResults`. But if this were to make a difference here, I would have expected that `SessionResults` works correctly and not the other way round. In any case, the actual error message should give a clue.

**Reply by 0J33** [reply]
Thanks for the help!

I figured out the problem.

After some testing with sending a message with different number of columns I discovered that Discord was having an issues every time I tried sending a message with more than 3 columns. I'm not sure why but when I changed it to to only display 3 columns it worked.

<img width="476" alt="sol" src="https://user-images.githubusercontent.com/90639992/193876997-a4037ce3-0118-49c4-8ed5-439d6bb44d22.png">
<img width="196" alt="3-r" src="https://user-images.githubusercontent.com/90639992/193877055-ad99f78d-1fe5-48c9-b0cf-7258411d9b6e.png">

**Reply by theOehrly** [reply]
@0J33 good to hear. Potentially, one of the columns that you are not sending now contains some special characters.

Alternatively, if this really is only about the number of columns, it may be caused by pandas shortening the datframe output using those dots that you see in the console as well. I'm not sure if this is done and how the maximum viewable size would be determinded in a case like this. Maybe you'd need to do some more explicit string formatting.
